tsdate/variational.py

```diff
@@ -651,55 +809,164 @@ def rescale(
+        skipped_muts = np.sum(np.isnan(self.mutation_posterior[:, 0]))
+        logger.info(f"Skipped {skipped_muts} mutations with invalid posteriors")
+        logger.info(f"Calculated mutation posteriors in {abs(muts_timing)} seconds")
+
+        singletons = self.mutation_blocks != tskit.NULL
+        switched_blocks = self.mutation_blocks[singletons]
+        switched_edges = np.where(
+            self.mutation_phase[singletons] < 0.5,
+            self.block_edges[switched_blocks, 1],
+            self.block_edges[switched_blocks, 0],
+        )
+        self.mutation_edges[singletons] = switched_edges
+        self.mutation_nodes[singletons] = self.edge_children[switched_edges]
+        switched = self.mutation_phase < 0.5
+        self.mutation_phase[switched] = 1 - self.mutation_phase[switched]
+        logger.info(f"Switched phase of {np.sum(switched)} singletons")
 
-        if rescale_intervals > 0:
+        if rescale_intervals > 0 and rescale_iterations > 0:
             rescale_timing = time.time()
-            self.rescale(
-                rescale_intervals=rescale_intervals, rescale_segsites=rescale_segsites
-            )
+            for _ in tqdm(
+                np.arange(rescale_iterations),
+                desc="Path Rescaling",
+                disable=not progress,
+            ):
+                self.rescale(
+                    rescale_intervals=rescale_intervals,
+                    rescale_segsites=rescale_segsites,
+                )
             rescale_timing -= time.time()
-            logging.info(f"Timescale rescaled in {abs(rescale_timing)} seconds")
+            logger.info(f"Timescale rescaled in {abs(rescale_timing)} seconds")
+
+    def node_moments(self):
+        """
+        Posterior mean and variance of node ages
+        """
+        alpha, beta = self.node_posterior.T
+        nodes_mn = np.ascontiguousarray(self.node_constraints[:, 0])
+        nodes_va = np.zeros(nodes_mn.size)
+        free = self.node_constraints[:, 0] != self.node_constraints[:, 1]
+        nodes_mn[free] = (alpha[free] + 1) / beta[free]
+        nodes_va[free] = nodes_mn[free] / beta[free]
+        return nodes_mn, nodes_va
+
+    def mutation_moments(self):
+        """
+        Posterior mean and variance of mutation ages
+        """
+        alpha, beta = self.mutation_posterior.T
+        muts_mn = np.full(alpha.size, np.nan)
+        muts_va = np.full(alpha.size, np.nan)
+        free = np.isfinite(alpha)
+        muts_mn[free] = (alpha[free] + 1) / beta[free]
+        muts_va[free] = muts_mn[free] / beta[free]
+        return muts_mn, muts_va
+
+    def mutation_mapping(self):
+        """
+        Map from mutations to edges and subtended node, using estimated singleton
+        phase (if singletons were unphased)
+        """
+        # TODO: should these be copies? Should members be readonly?
+        return self.mutation_edges, self.mutation_nodes
+
+
+# def date(
```

Comment:
should we just remove this?

Reply:
I use this for debugging (when I need the expectationpropagation object) so would rather keep it around for the moment